DataFormats/Detectors/FIT/FT0/include/DataFormatsFT0/RawEventData.h

```diff
@@ -0,0 +1,213 @@
+             uint32_t(h->detectorField), uint32_t(h->par), uint32_t(h->stop), uint32_t(h->pageCnt));
+    }
+  }
+  std::vector<char> to_vector()
+  {
+    constexpr int CRUWordSize = 16;
+    const char padding[CRUWordSize] = {0};
+
+    std::vector<char> result(size() * CRUWordSize);
+    char* out = result.data();
+    std::memcpy(out, &mEventHeader, EventHeader::PayloadSize);
+    out += EventHeader::PayloadSize;
+    LOG(DEBUG) << "write header for " << (int)mEventHeader.nGBTWords << " orbit " << int(mEventHeader.orbit) << " bc " << int(mEventHeader.bc);
+    if (mIsPadded) {
+      out += CRUWordSize - EventHeader::PayloadSize;
+      LOG(INFO) << "padding header " << CRUWordSize - EventHeader::PayloadSize;
+    }
+    for (int i = 0; i < mEventHeader.nGBTWords; ++i) {
+      std::memcpy(out, &mEventData[2 * i], EventData::PayloadSize);
+      out += EventData::PayloadSize;
+      std::memcpy(out, &mEventData[2 * i + 1], EventData::PayloadSize);
+      out += EventData::PayloadSize;
+      if (mIsPadded) {
+        out += CRUWordSize - 2 * EventData::PayloadSize;
+        LOG(DEBUG) << "padding data " << CRUWordSize - 2 * EventData::PayloadSize;
+      }
+    }
+    return result;
+  }
+  void setIsPadded(bool isPadding128)
+  {
+    mIsPadded = isPadding128;
+  }
+
+ public:
+  EventHeader mEventHeader;
+  EventData mEventData[Nchannels_PM];
+  bool mIsPadded = true;
+  /////////////////////////////////////////////////
+  ClassDefNV(RawEventData, 1);
+};
+std::ostream& operator<<(std::ostream& stream, const RawEventData& data);
+
+class DataPageWriter
+{
+  std::vector<char> mBuffer;
+  int mNpacketsInBuffer = 0;
+  std::vector<std::vector<char>> mPages;
+  std::vector<int> mNpackets;
+  static constexpr int MAX_PAGE_SIZE = 8192;
+
+ public:
+  o2::header::RAWDataHeader mRDH;
+  void flush(std::ostream& str)
+  {
+    writePage();
+    mRDH.stop = 0;
+    for (int page = 0; page < int(mPages.size()); ++page) {
+      mRDH.memorySize = mPages[page].size() + mRDH.headerSize;
+      mRDH.offsetToNext = mRDH.memorySize;
+      mRDH.packetCounter = mNpackets[page];
+      RawEventData::printRDH(&mRDH);
+      str.write(reinterpret_cast<const char*>(&mRDH), sizeof(mRDH));
+      str.write(mPages[page].data(), mPages[page].size());
+      mRDH.pageCnt++;
+    }
+    if (!mPages.empty()) {
+      mRDH.memorySize = mRDH.headerSize;
+      mRDH.offsetToNext = mRDH.memorySize;
+      mRDH.stop = 1;
+      RawEventData::printRDH(&mRDH);
+      str.write(reinterpret_cast<const char*>(&mRDH), sizeof(mRDH));
+      mRDH.pageCnt++;
+      mPages.clear();
+      mNpackets.clear();
+    }
+  }
+
+  void writePage()
+  {
+    if (mBuffer.size() == 0)
+      return;
+    mPages.emplace_back(std::move(mBuffer));
+    mNpackets.push_back(mNpacketsInBuffer);
+    mNpacketsInBuffer = 0;
+    mBuffer.clear();
+  }
+
+  void write(std::vector<char> const& new_data)
+  {
+    if (mBuffer.size() + new_data.size() + mRDH.headerSize > MAX_PAGE_SIZE)
```

Comment:
should not you first complete the existing page till MAX_PAGE_SIZE and only then open a new page (i.e. you need to truncate the new_data between the old and new pages).

Reply:
really? I think i have to close  current page and move new data to next one. Page size will be less than 8Kb. But as I understand from Filippo slides offset should be 8K any case

Reply:
The offset should be RDH_size + stored_payload_size and <= MAX_PAGE_SIZE. But here you are carrying whole new_data to the new page if its size + current ``RDH_size + stored_payload_size`` exceeds 8K. Instead, one should split the new_data between the current and new pages. At least this is what ITS is going to implement, I assume the same is true for other detectors. Could you check with your HW experts if you are going to split the single trigger data between 2 pages?

Reply:
What is done here is : if data from 1 link&1BC (stored_payload_size) + RDH are >8K I just close previous page and write this data to new one. Are HW people said me this.

Reply:
If you are sure that is what will be done in HW, then OK (ITS will try to complete the 8K then create a new page).
Then the main problem remaining is the header: if your HW people cannot arrange its members in computer-readable format, then we need change it to a class with private members, split some fields (orbit) and add getters to assemble them.

Reply:
they suggested structure like this for header
 struct EventHeaderRaw {//no padding
    union {
      uint64_t word[2];
      struct	{
	uint64_t bc:12,
	  orbit:32,
	  reservedField1:20,
	  reservedField2:8,
	  nGBTWords:4,
	  startDescriptor:4,
	  reservedField3:48;
      };
    };
I  checked my code with this structure - it works
![T0ACdigrec](https://user-images.githubusercontent.com/20040756/69645913-3aa83700-1067-11ea-95a0-6b66cbdade82.png)

Reply:
Yes, this will work also. But currently only 56 bits are used, why not putting all of them in the 1st 64b word, i.e.
````
struct EventHeaderRaw {//no padding
  union {
    uint64_t word[2];
    struct {
      uint16_t bc : 12;
      uint32_t orbit : 32;
      uint8_t nGBTWords : 4;
      uint8_t startDescriptor : 4;
      uint16_t reservedField1 : 12;
      uint64_t reservedField2 : 16;
   };
 };
};
````

Reply:
OK

Reply:
I think it is not my code crashes in build/O2/o2

Reply:
no, this is CI problem.

Reply:
Should I add HBF now? Or this can be done later?